sklearn/utils/class_weight.py

```diff
@@ -41,12 +41,14 @@ def compute_class_weight(class_weight, classes, y):
     # Import error caused by circular imports.
     from ..preprocessing import LabelEncoder
 
+    _dtype = y.dtype
+
     if set(y) - set(classes):
         raise ValueError("classes should include all valid labels that can "
                          "be in y")
     if class_weight is None or len(class_weight) == 0:
         # uniform class weights
-        weight = np.ones(classes.shape[0], dtype=np.float64, order='C')
+        weight = np.ones(classes.shape[0], dtype=_dtype, order='C')
```

Comment:
I think that I disagree. You will have an operation between `sample_weight` and `class_weight` for instance [there](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/logistic.py#L622). Because you make a multiplication in-place, I agree that `sample_weight` will be the expected type. However, `class_weight` should not always be casted to `np.float64` if the type of `sample_weight` is `np.float32`. It will imply a conversion when making the array conversion.

What I mean is something like that:

```python
x = np.random.random((1000000, 1)).astype(np.float32)
y = np.random.random((1000000, 1)).astype(np.float32)
%timeit x * y
```
```
807 µs ± 17.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

```python
x = np.random.random((1000000, 1)).astype(np.float32)
y = np.random.random((1000000, 1)).astype(np.float64)
%timeit x * y
```

```
1.83 ms ± 8.57 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Does it make sense?